### PR TITLE
Build package initial refactor for multistage 

### PR DESCRIPTION
**Pull request by ikaneshiro**
**Description of the Pull Request (PR):**

Introduces the `stage` struct to the build package. This encapsulates the data needed to perform the building of a stage. Allowing the `Build` struct to hold a slice of `stage`s to be build in order when `deffile` syntax is expanded to allow for multiple `bootstraps` in a single file.

Introduces the `Config` struct to the build package. This contains build information as well as a `types.Options` struct to pass on information to newly constructed `bundles`

Moved `libraryURL` and `authToken` to `types.Options` struct since this data should exist in the `Bundle`

Created a `metadata.go` file to hold helper functions that assist the insertion of metadata to a `Bundle` during the build process.

Minor additional refinements.

**This fixes or addresses the following GitHub issues:**

- Related to #2518

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by ArangoGutierrez** [comment]
Heya @ikaneshiro , I know this is still tagged as WIP, but,  pls don't forget about testing ;) 
Cheers

File: internal/pkg/build/metadata.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (WestleyK) [inline internal/pkg/build/metadata.go]: 2019?

```suggestion
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

Or:

```suggestion
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
```

File: internal/pkg/build/sources/conveyorPacker_library.go [hunk 2]
```diff
@@ -31,6 +29,9 @@ func (cp *LibraryConveyorPacker) Get(b *types.Bundle) (err error) {
 
 	cp.b = b
 
+	libraryURL := b.Opts.LibraryURL
```
Comment (WestleyK) [inline internal/pkg/build/sources/conveyorPacker_library.go:32]: I would update the copyright in this file, `2018-2019`.

File: internal/pkg/build/stage.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (WestleyK) [inline internal/pkg/build/stage.go]: Update copyright:

```suggestion
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

Or:

```suggestion
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
```

File: internal/pkg/build/stage.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+	"github.com/sylabs/singularity/pkg/build/types"
+)
+
+// Stage ...
```
Comment (WestleyK) [inline internal/pkg/build/stage.go]: Is there more to this comment?

File: internal/pkg/libexec/pull.go [hunk 5]
```diff
@@ -40,7 +40,7 @@ func PullShubImage(filePath, shubRef string, force, noHTTPS bool) {
 
 // PullOciImage pulls an OCI image to a sif
 func PullOciImage(path, uri string, opts types.Options) {
-	b, err := build.NewBuild(uri, path, "sif", "", "", opts)
+	b, err := build.NewBuild(uri, build.Config{Dest: path, Format: "sif", Opts: opts})
```
Comment (WestleyK) [inline internal/pkg/libexec/pull.go:43]: I would update the copyright in this file: `2018-2019`. *I cant comment that line, github wont let me.*

File: pkg/build/types/bundle.go [hunk 6]
```diff
@@ -41,6 +41,16 @@ type Bundle struct {
 
 // Options ...
 type Options struct {
+	// sections are the parts of the definition to run during the build
```
Comment (WestleyK) [inline pkg/build/types/bundle.go:44]: Update copyright in this file too.

**Review by WestleyK — COMMENTED** [review]
Minor suggestions for the copyright year, otherwise looks good!

File: pkg/build/types/bundle.go (hunk 6 above)
Comment (bauerm97) [inline pkg/build/types/bundle.go:44]: Can we actually write a comment on this type please?

File: internal/pkg/build/build.go [hunk 7]
```diff
@@ -41,90 +37,112 @@ import (
 // 		Execute all of a definition using AllSections()
 // 		And finally call Assemble() to create our container image
 type Build struct {
+	// stages of the build
+	stages []stage
+	// Config
+	Conf Config
+}
+
+// Config ...
```
Comment (bauerm97) [inline internal/pkg/build/build.go]: And not to rain on the parade, but let's comment this as well

File: internal/pkg/build/build.go [hunk 8]
```diff
@@ -41,90 +37,112 @@ import (
...
+type Config struct {
 	// dest is the location for container after build is complete
-	dest string
+	Dest string
```
Comment (bauerm97) [inline internal/pkg/build/build.go:49]: The comment on this field should also be updated to match the case

File: internal/pkg/build/build.go [hunk 9]
```diff
@@ -41,90 +37,112 @@ import (
...
 	// format is the format of built container, e.g., SIF, sandbox
-	format string
-	// c Gets and Packs data needed to build a container into a Bundle from various sources
-	c ConveyorPacker
-	// a Assembles a container from the information stored in a Bundle into various formats
-	a Assembler
-	// b is an intermediate structure that encapsulates all information for the container, e.g., metadata, filesystems
-	b *types.Bundle
+	Format string
+	// noCleanUp allows a user to prevent a bundle from being cleaned up after a failed build
```
Comment (bauerm97) [inline internal/pkg/build/build.go]: Same here

File: internal/pkg/build/build.go [hunk 10]
```diff
@@ -41,90 +37,112 @@ import (
...
+	// useful for debugging
+	NoCleanUp bool
+	// Options for bundles
```
Comment (bauerm97) [inline internal/pkg/build/build.go]: And here

File: internal/pkg/build/build.go [hunk 11]
```diff
@@ -41,90 +37,112 @@ import (
...
+	Dest string
 	// format is the format of built container, e.g., SIF, sandbox
-	format string
-	// c Gets and Packs data needed to build a container into a Bundle from various sources
-	c ConveyorPacker
-	// a Assembles a container from the information stored in a Bundle into various formats
-	a Assembler
-	// b is an intermediate structure that encapsulates all information for the container, e.g., metadata, filesystems
-	b *types.Bundle
+	Format string
```
Comment (bauerm97) [inline internal/pkg/build/build.go:51]: And here

File: internal/pkg/build/build.go [hunk 12]
```diff
@@ -41,90 +37,112 @@ import (
 // 		Execute all of a definition using AllSections()
 // 		And finally call Assemble() to create our container image
 type Build struct {
+	// stages of the build
+	stages []stage
+	// Config
+	Conf Config
```
Comment (bauerm97) [inline internal/pkg/build/build.go:43]:
```
// Conf [...]
```

**Review by bauerm97 — CHANGES_REQUESTED** [review]

**Review by bauerm97 — APPROVED** [review]
I think this looks good for a merge. I do think that for this particular commit (since it's touching a lot of code affecting the build process) we should make sure the integration tests are passing not just the unit tests

**Pull request merged**